File: smartmoneyserver/src/main/java/fpt/aptech/server/dto/transaction/report/TransactionReportResponse.java
package fpt.aptech.server.dto.transaction.report;

import lombok.Builder;

import java.math.BigDecimal;

/**
 * DTO trả về dữ liệu tổng quan cho màn hình "Báo cáo" (Tab Báo cáo).
 * Bao gồm:
 * 1. Phần Header: Tổng quan tài chính (Vào/Ra/Ròng) + Số dư đầu/cuối kỳ.
 * 2. Phần Thống kê: Nhóm Nợ/Vay/Khác (trả về số tiền để hiển thị ở footer).
 */
@Builder
public record TransactionReportResponse(
    // --- Số dư ---
    BigDecimal openingBalance, // Số dư đầu kỳ
    BigDecimal closingBalance, // Số dư cuối kỳ

    // --- Tổng quan trong kỳ ---
    BigDecimal totalIncome,  // Tổng tiền thu vào
    BigDecimal totalExpense, // Tổng tiền chi ra
    BigDecimal netIncome,    // Thu nhập ròng (Thu - Chi)

    // --- 3. Nhóm Nợ/Vay/Khác (Phần Footer - Click để xem chi tiết) ---
    // Trả về số tiền để hiển thị theo nhóm chuyên biệt (dùng cho footer trên UI)
    // debtAmount: Tổng tiền của tất cả giao dịch thuộc category "Đi vay" (DEBT_BORROWING = 20)
    BigDecimal debtAmount,
    // loanAmount: Tổng tiền của tất cả giao dịch thuộc category "Cho vay" (DEBT_LENDING = 19)
    BigDecimal loanAmount,
    // otherAmount: Số tiền còn lại = tổng(Thu nợ DEBT_COLLECTION = 21) - tổng(Trả nợ DEBT_REPAYMENT = 22)
    // Ví dụ: Thu nợ 22.222 - Trả nợ 11.111 => otherAmount = 11.111
    BigDecimal otherAmount
)
{}

File: APP MOBILE/smart_money/lib/modules/transaction/models/report/transaction_report_response.dart
// modules/transaction/models/report/transaction_report_response.dart
// Map đúng với TransactionReportResponse.java của Spring Boot
// Tổng quan báo cáo tài chính theo kỳ (REPORT)

import '../view/transaction_response.dart';

class TransactionReportResponse {
  final double openingBalance;   // Số dư đầu kỳ
  final double closingBalance;   // Số dư cuối kỳ
  final double totalIncome;      // Tổng tiền thu vào
  final double totalExpense;     // Tổng tiền chi ra
  final double netIncome;        // Thu nhập ròng (Thu - Chi)
  final int debtTransactionCount;  // Số giao dịch Nợ (Đi vay, Trả nợ)
  final int loanTransactionCount;  // Số giao dịch Cho vay (Cho vay, Thu nợ)
  final List<TransactionResponse>? transactions;

  const TransactionReportResponse({
    required this.openingBalance,
    required this.closingBalance,
    required this.totalIncome,
    required this.totalExpense,
    required this.netIncome,
    required this.debtTransactionCount,
    required this.loanTransactionCount,
    this.transactions,
  });

  factory TransactionReportResponse.fromJson(Map<String, dynamic> json) {
    return TransactionReportResponse(
      openingBalance: (json['openingBalance'] as num).toDouble(),
      closingBalance: (json['closingBalance'] as num).toDouble(),
      totalIncome: (json['totalIncome'] as num).toDouble(),
      totalExpense: (json['totalExpense'] as num).toDouble(),
      netIncome: (json['netIncome'] as num).toDouble(),
      debtTransactionCount: json['debtTransactionCount'] as int? ?? 0,
      loanTransactionCount: json['loanTransactionCount'] as int? ?? 0,
      transactions: json['transactions'] != null
          ? (json['transactions'] as List<dynamic>)
              .map((e) => TransactionResponse.fromJson(e as Map<String, dynamic>))
              .toList()
          : null,
    );
  }
}

